### PR TITLE
Fix locating program in exec watch mode

### PR DESCRIPTION
**Pull request by gridbugs**
This fixes a bug where running an executable from the current project
with `dune exec --watch` would be unable to find the executable unless
the command was run from the project's root directory.

The problem was introduced when we started setting the cwd of processes
spawned by exec in watch mode to the user's current directory
(https://github.com/ocaml/dune/pull/10262). If the program argument to
exec refers to a file to be built in the current project (such as an
executable implemented in the current project) then the path to the
executable to spawn will be a path inside the _build directory relative
to the project root. Since the cwd of the new process was set to the
user's current directory, this relative path was being resolved within
the current directory, whereas it should have been resolved relative to
the project root.

The fix was to convert relative paths into absolute paths relative to
the project root (this was already being done for exec when not in watch
mode).

### CODE REVIEW

File: test/blackbox-tests/test-cases/exec-watch/exec-watch-multi-levels.t/run.t [hunk 1]
```diff
@@ -0,0 +1,34 @@
+Test the behaviour of exec watch mode when the current directory is not the
```
Comment (emillon) [inline test/blackbox-tests/test-cases/exec-watch/exec-watch-multi-levels.t/run.t:1]: This is a repro for this issue, right? can you land it first in its own PR?

File: bin/import.ml [hunk 2]
```diff
@@ -239,14 +239,16 @@ module Scheduler = struct
...
 
-let restore_cwd_and_execve (common : Common.t) prog argv env =
-  let prog =
-    if Filename.is_relative prog
-    then (
-      let root = Common.root common in
-      Filename.concat root.dir prog)
-    else prog
-  in
+let resolve_relative_path_within_root_dir (common : Common.t) path =
```
Comment (anmonteiro) [inline bin/import.ml:242]: you're threading `Common.t` all the way here just to get `Common.root`. how about we just take `~root` instead?

**Comment by gridbugs** [comment]
> what happens with `cd path && dune exec -- echo hello`?
> 
> IIUC, `echo` would pass the `Filename.is_relative` check and get a prefix when it shouldn't?

`dune exec` converts paths with no slashes into absolute paths by looking them up in PATH, so any program names will be represented by their absolute path here.

File: test/blackbox-tests/test-cases/exec-watch/exec-watch-multi-levels.t/run.t (hunk 1 above)
Comment (gridbugs) [reply test/blackbox-tests/test-cases/exec-watch/exec-watch-multi-levels.t/run.t:1]: Made a new pr here: https://github.com/ocaml/dune/pull/10391

File: bin/import.ml (hunk 2 above)
Comment (gridbugs) [reply bin/import.ml]: fixed

**Comment by gridbugs** [comment]
Just making sure that we don't forget to merge this. Are any other changes necessary?

**Comment by rgrinberg** [comment]
Fixed in another PR
